aten/src/ATen/core/copy_on_write.cpp

```diff
@@ -0,0 +1,19 @@
+#include <ATen/core/copy_on_write.h>
+
+#include <ATen/core/TensorBase.h>
+
+namespace at {
+
+auto simulate_materialize_copy_on_write(TensorBase const& tensor) -> void {
+  if (!tensor.has_storage()) {
+    return;
+  }
+
+  c10::TensorImpl* tensor_impl = tensor.unsafeGetTensorImpl();
+  if (tensor_impl == nullptr) {
```

Comment:
This feels like it should be an assert failure instead

Reply:
Is this even possible? Given that TensorBase's member uses UndefinedTensorImpl as its NullType?

Reply:
done